src/boot/efi/efi-string.c

```diff
@@ -0,0 +1,44 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "efi-string.h"
+
+/* String functions for both char and char16_t that should behave the same way as their respective
+ * counterpart in userspace. Where it makes sense, these accept NULL and do something sensible whereas
+ * userspace does not allow for this (strlen8(NULL) returns 0 like strlen_ptr(NULL) for example). To make it
+ * easier to tell in code which kind of string they work on, we use 8/16 suffixes. This also makes is easier
+ * to unit test them.
+ *
+ * To avoid repetition and because char/char16_t string functions are fundamentally the same, we implement
+ * them with macros. Note that these macros are to be only used within their respective function they
+ * implement and therefore do not care about parameter hygiene and will also use the function parameters
+ * without declaring them in the macro params. */
+
+#define STRNLEN_U                       \
```

Comment:
Of course, this is used only in defining functions below, but I prefer to declare macros in a function like form:
```
#define STRLEN(s) \
({ \
    typeof _s = (s); \
    if (!_s) \
        return 0; \
    ...
})
```

Reply:
I don't think it adds much to the table. The `return` statements would still make this not a function-like macro. And reordering the control flow because statement expressions can only return something on the last line would make this way too ugly imho.